### PR TITLE
lazyasd

### PR DESCRIPTION
**Pull request by scopatz**
New package for speeding up module imports.

### CODE REVIEW

**Comment by jakirkham** [comment]
Seems to get the following error on the Python 2.7 build.

```
Compiling /opt/conda/envs/_build/lib/python2.7/site-packages/lazyasd.py ...
  File "/opt/conda/envs/_build/lib/python2.7/site-packages/lazyasd.py", line 68
    yield from obj
```

Is this compatible with Python 2.7?

**Comment by scopatz** [comment]
Ahh yeah, it is probably not py2k compliant yet.

File: recipes/lazyasd/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+  run:
+    - python
+
+test:
+  imports:
+    - lazyasd
+  requires:
+    - pytest
+  commands:
+    - py.test
```
Comment (jakirkham) [inline recipes/lazyasd/meta.yaml]: We might need to switch to `SRC_DIR` first. Though it is a bit of kludgy hack, it does work and doesn't seem to cause any harm here.

Comment (scopatz) [inline recipes/lazyasd/meta.yaml]: OK, I have tried that.

**Comment by jakirkham** [comment]
Seems we are having some downloading issues. Have raised it upstream ( https://github.com/Anaconda-Platform/support/issues/49 ).

**Comment by msarahan** [comment]
I looked briefly at what it would take to have py2k support.  The yield from's are easy to convert, but I haven't found an equivalent for importlib.utils.resolve_name.

File: recipes/lazyasd/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  run:
+    - python
+
+test:
+  imports:
+    - lazyasd
+  requires:
+    - pytest
+  commands:
+    - cd $SRC_DIR && py.test
```
Comment (jakirkham) [inline recipes/lazyasd/meta.yaml]: Please change to the following

```
    - cd $SRC_DIR   # [unix]
    - cd %SRC_DIR%  # [win]
    - py.test
```

**Comment by jakirkham** [comment]
Seems downloads are working again. Restarted the CIs and everything passes except Windows (no surprise there). Have noted the changes above that should make this pass.

**Comment by jakirkham** [comment]
> I looked briefly at what it would take to have py2k support.  The yield from's are easy to convert, but I haven't found an equivalent for importlib.utils.resolve_name.

To handle `importlib` in Python 2.x, I think we just need [`importlib2`]( https://pypi.python.org/pypi/importlib2 ).

**Comment by msarahan** [comment]
Yes, I think you're right about importlib2.  Google failed me completely on that.  "importlib python backport" "importlib.utils python 2 backport" and the like.  I'll submit a PR later for this.

File: recipes/lazyasd/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,56 @@
+# Jinja variables help maintain the recipe as you'll update the version only here.
```
Comment (jakirkham) [inline recipes/lazyasd/meta.yaml]: Please trim this comment.

File: recipes/lazyasd/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+{% set sha256 = "3be134452b5933918dbbf32809999b2f6786fc17dd3a0dcca328afcf4370e292" %}
+
+package:
+  name: lazyasd
+  version: {{ version }}
+
+source:
+  fn: lazyasd-{{ version }}.tar.gz
+  url: https://github.com/xonsh/lazyasd/archive/{{ version }}.tar.gz
+  # md5 is OK but sha256 is desirable.
```
Comment (jakirkham) [inline recipes/lazyasd/meta.yaml]: Please trim this comment.

File: recipes/lazyasd/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+  sha256: {{ sha256 }}
+
+build:
+  number: 0
+  # If the installation is complex, or different between Unix and Windows, use separate bld.bat and build.sh files instead of this key.
```
Comment (jakirkham) [inline recipes/lazyasd/meta.yaml]: Please trim this comment.

File: recipes/lazyasd/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+  script: python setup.py install --single-version-externally-managed --record record.txt
+  skip: True  # [py2k]
+
+requirements:
+  build:
+    - python
+    # When setuptools is available add the `--single-version-externally-managed --record record.txt` above.
```
Comment (jakirkham) [inline recipes/lazyasd/meta.yaml]: Please trim this comment.

**Comment by scopatz** [comment]
Thanks for the cd hint @jakirkham! Also, happy to have Py2k support for this.  The only gotcha is that the implementation needs to be amalgamate friendly (no try-excepting ImportErrors).  Also, it would be great if it could be based solely on the standard library (though for Python 2, I don't personally).  

But really resolve_name is easy enough to implement.  Here is the whole source code from the standard library:

```python
def resolve_name(name, package):
    """Resolve a relative module name to an absolute one."""
    if not name.startswith('.'):
        return name
    elif not package:
        raise ValueError('{!r} is not a relative name '
                         '(no leading dot)'.format(name))
    level = 0
    for character in name:
        if character != '.':
            break
        level += 1
    return _resolve_name(name[level:], package, level)
```

File: recipes/lazyasd/meta.yaml [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    - cd $SRC_DIR   # [unix]
+    - cd %SRC_DIR%  # [win]
+    - py.test
+
+about:
+  home: https://github.com/xonsh/lazyasd/blob/master/README.rst#lazyasd
+  license: BSD 3-clause
+  summary: 'Lazy & self-destructive tools for speeding up module imports'
+
+  # The remaining entries in this section are optional, but recommended
```
Comment (jakirkham) [inline recipes/lazyasd/meta.yaml]: Please trim this comment.

**Comment by jakirkham** [comment]
Some comments leftover from the example should be cut. Otherwise LGTM.

**Comment by scopatz** [comment]
Which I guess hides a C-func, but my point is that with level info too this is not hard to implement.  Or we could just require that Python 2 uses absolute imports.

**Comment by scopatz** [comment]
> Some comments leftover from the example should be cut. Otherwise LGTM.

It seems like there should be a comment in the example about removing the comments :)  And also done.

**Comment by jakirkham** [comment]
Thanks @scopatz.

**Comment by scopatz** [comment]
Thanks a ton!

**Pull request merged**